### PR TITLE
Sprint-1 and Sprint-2 features

### PR DESCRIPTION
**Pull request by buraktomruk**
Requested changes for sprint-1 are also added in this PR.

### CODE REVIEW

File: react-app/src/App.test.tsx [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import App from './App';
+
+it('renders without crashing', () => {
```
Comment (eneufeld) [inline react-app/src/App.test.tsx:5]: it probably makes sense to add some tests?

File: react-app/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+      switch (message.data.command) {
+        case 'setTutorials':
+          setTutorials(message.data.tutorials);
+          break;
+      }
+    });
+  });
+
+  const createTutorialList = () => {
```
Comment (eneufeld) [inline react-app/src/App.tsx:33]: if you want to pull this out for readability, add a new functional component, it is cleaner

File: react-app/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          {createTutorialList()}
+          </p></p>
+        :
+        <SnackbarProvider maxSnack={3}>
+          <TutorialOverview tutorial={tutorials[selectedTutorial]} />
+        </SnackbarProvider>
+      }
+    </div>
+
+
```
Comment (eneufeld) [inline react-app/src/App.tsx]: remove unneeded empty lines

File: react-app/src/components/Command.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+export const Command = (props: CommandProps) => {
+
+  return (
+    <div className="Box-margin">
+      <Button onClick={() => {
+        VSCodeAPI.postMessage({commands: props.button.button.commands, ids: [], exerciseFolder: props.exerciseFolder});
+      }} variant="contained" style={{
+        margin: "10px 10px 10px 10px",
```
Comment (eneufeld) [inline react-app/src/components/Command.tsx]: why use material ui if you completely resyle the button?

File: react-app/src/components/StepperComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+import {VSCodeAPI} from '../VSCodeAPI';
+import SettingsIcon from '@material-ui/icons/Settings';
+import {TestFeedbackDialog} from './TestFeedbackDialog';
+import {Exercise, Tutorial} from '../../../schema/tutorial';
+import {vsTheme} from '../VsTheme';
+import MuiDialogTitle from '@material-ui/core/DialogTitle';
+import CloseIcon from '@material-ui/icons/Close';
+import MuiDialogContent from '@material-ui/core/DialogContent';
+
+const useStyles = makeStyles((theme: Theme) =>
```
Comment (eneufeld) [inline react-app/src/components/StepperComponent.tsx:29]: I suggest to either completely get rid of material ui or at least do a custom theme were you redefine everything

File: react-app/src/components/StepperComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+const getSteps = (tutorialExercises: Array<Exercise> | undefined) => {
+  return tutorialExercises && tutorialExercises.map((tutorial) => tutorial.title);
+}
+
+interface StepperComponentProps {
+  tutorial: Tutorial;
+  startStep: number;
+}
+
+const StepperComponent = (props: StepperComponentProps) => {
```
Comment (eneufeld) [inline react-app/src/components/StepperComponent.tsx:87]: You should split the rendering in some sub components, it is hard to understand what is going on

File: react-app/src/components/TestFeedbackDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        </MuiDialogTitle>
+        <MuiDialogContent dividers>
+            {props.test != null && createTestParagraph(props.test, props.exerciseFolder)}
+        </MuiDialogContent>
+    </Dialog></div >;
+};
+
+const createTestParagraph = (tests: Array<TestParagraph>, exerciseFolder: String) => {
+    return tests.map((testParagraph) => {
+        return <div><Grid
```
Comment (eneufeld) [inline react-app/src/components/TestFeedbackDialog.tsx:75]: if you just need an element to wrap, you can use: `<>` instead

File: react-app/src/components/TutorialOverview.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+const TutorialOverview = (props: TutorialOverviewProps) => {
+    const classes = useStyles();
+    const [activeStep, setActiveStep] = useState(-1);
+
+    return (
+        activeStep < 0 ?
+            <div className={classes.root}>
+                <Stepper orientation="vertical" style={{backgroundColor: "transparent"}}>
```
Comment (eneufeld) [inline react-app/src/components/TutorialOverview.tsx]: Pull out as component

File: react-app/src/components/TutorialOverview.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            textAlign: "left",
+        },
+    }),
+);
+
+interface TutorialOverviewProps {
+    tutorial: Tutorial;
+}
+
+const TutorialOverview = (props: TutorialOverviewProps) => {
```
Comment (eneufeld) [inline react-app/src/components/TutorialOverview.tsx]: you can deconstruct in line:
```suggestion
const TutorialOverview = ({tutorial}:TutorialOverviewProps) => {
```

File: src/Commands/AddImportsCommand.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import * as vscode from 'vscode';
+import {AutomaticImport} from '../../schema/tutorial';
+const fs = require('fs');
+const path = require('path');
+
+const ADDIMPORTSCOMMAND: vscode.Disposable = vscode.commands.registerCommand('theiatutorialextension.addImports', (autoImportData: AutomaticImport) => {
+
+    const workspaceFolder: string = vscode.workspace.rootPath || '~';
+
+    let filepath = path.join(workspaceFolder, autoImportData.automaticImport.path);
```
Comment (eneufeld) [inline src/Commands/AddImportsCommand.ts]: why `let`? can't we use const?

File: src/Commands/AddImportsCommand.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    let content = fs.readFileSync(filepath);
```
Comment (eneufeld) [inline src/Commands/AddImportsCommand.ts:22]: same as above

File: src/Commands/CloneRepoCommand.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0, or the MIT License which is
+ * available at https://opensource.org/licenses/MIT.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR MIT
+ ********************************************************************************/
+import * as vscode from 'vscode';
+import cloneRepo from '../Functions/cloneRepo';
+
+const CLONEREPOCOMMAND: vscode.Disposable = vscode.commands.registerCommand('theiatutorialextension.cloneRepo', () => {
```
Comment (eneufeld) [inline src/Commands/CloneRepoCommand.ts]: why is it all uppercase is this some kind of convention for commands?

**Review by eneufeld — CHANGES_REQUESTED** [review]
Is the `helloworld` file needed?
I would that you suggest setting up the repo with lint rules and formatting as defined here: https://github.com/sdirix/react-app-starter/ so configure lint and prettier
This will increase the readability a lot.
Please also try to create small components, they are easier to refactor and understand

**Comment by eneufeld** [comment]
So overall the commit is ok, I suggest to fix at least the formatting by configuring prettier and eslint .
Furthermore it would be great if you could check that all files have headers (should be handled by eslint ideally). 
After this 2 points this PR can be merged so that an initial version is in.

After that some more cleanup will be needed:
- there are some `let` uses where you could use `const` imho
- split react components to smaller ones so that the readability is improved
- use react components instead of functions to create dom elements 

On the long run you should discuss whether material-ui is the best choice here.

**Comment by eneufeld** [comment]
Ok I talked with Jonas, so I suggest to do the following:
- check all files and add missing headers -> merge afterwards
- create individual issues for all open points
Thank you!

File: src/Commands/LoadConfigCommand.ts [hunk 13]
```diff
@@ -30,16 +30,16 @@ const processTutorials = (tutorials: Array<Tutorial>): void => {
...
-                if (tutorial.exercises[i].solve !== null) {
+                if (tutorial.exercises[i].solve != null && tutorial.exercises[i].cleanState != null) {
                     tutorial.exercises[i].solve = tutorial.exercises[i].cleanState?.concat(tutorial.exercises[i].solve!);
                 }
-                if (tutorial.exercises[i].buildExercise === null && i !== 0) {
+                if (tutorial.exercises[i].buildExercise == null && i !== 0) {
                     tutorial.exercises[i].buildExercise = tutorial.exercises[i - 1].solve;
                 }
-                if (tutorial.exercises[i].checkStartState === null && i !== 0) {
+                if (tutorial.exercises[i].checkStartState == null && i !== 0) {
```
Comment (eneufeld) [inline src/Commands/LoadConfigCommand.ts:42]: why replace the `===` with `==` ?

**Comment by eneufeld** [comment]
Looks good overall. Please make sure to sign the eclipse eca.
Please fix missing headers, after that we can merge.
Please open issues for all open review comments.

**Comment by buraktomruk** [comment]
The new PR is #6 I can close this

**Pull request closed without merging**